### PR TITLE
SD3 and Flux support

### PR DESCRIPTION
**Pull request by IlyasMoutawwakil**
# What does this PR do?

In this PR we add support of SD3 and Flux Diffusion Pipelines, for onnx export and onnxruntime inference.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/exporters/onnx/model_configs.py [hunk 1]
```diff
@@ -1160,6 +1154,171 @@ def outputs(self) -> Dict[str, Dict[int, str]]:
...
+        }
+
+    @property
+    def outputs(self):
+        return {
+            "last_hidden_state": {0: "batch_size", 1: "sequence_length"},
+        }
+
+
+class DummyTransformerTimestpsInputGenerator(DummyTimestepInputGenerator):
```
Comment (JingyaHuang) [inline optimum/exporters/onnx/model_configs.py:1171]: Why don't you put these dummy input generators under the inputs generator file?

Comment (IlyasMoutawwakil) [reply optimum/exporters/onnx/model_configs.py:1171]: I thought that's what we do when  the dummy input generator is model-specific, will move them 👍

**Review by JingyaHuang — APPROVED** [review]
LGTM as far as we get the CIs green. Thanks for adding the support @IlyasMoutawwakil !

File: optimum/exporters/tasks.py [hunk 2]
```diff
@@ -335,15 +335,27 @@ class TasksManager:
...
-        "unet": supported_tasks_mapping(
+        "flux-transformer-2d-model": supported_tasks_mapping(
+            "semantic-segmentation",
+            onnx="FluxTransformerOnnxConfig",
+        ),
+        "sd3-transformer-2d-model": supported_tasks_mapping(
+            "semantic-segmentation",
+            onnx="SD3TransformerOnnxConfig",
+        ),
+        "unet-2d-condition": supported_tasks_mapping(
```
Comment (IlyasMoutawwakil) [inline optimum/exporters/tasks.py:358]: @JingyaHuang we change model types here, normally it shouldn't be breaking as these model types are only used in the lib internally by the exporter when exporter diffusion pipelines.

Comment (IlyasMoutawwakil) [reply optimum/exporters/tasks.py:358]: tell me if this breaks something I'm not aware of 😄

**Comment by kahmed10** [comment]
How would this work with stabilityai/stable-diffusion-3-medium on huggingface?

Even after specifying the framework as pytorch, it's not able infer the correct library for the models:
```
optimum-cli export onnx --model stabilityai/stable-diffusion-3-medium --framework pt  models/sd3-onnx
  File "/home/kahmed10/.local/bin/optimum-cli", line 8, in <module>
    sys.exit(main())
  File "/home/kahmed10/.local/lib/python3.10/site-packages/optimum/commands/optimum_cli.py", line 208, in main
    service.run()
  File "/home/kahmed10/.local/lib/python3.10/site-packages/optimum/commands/export/onnx.py", line 265, in run
    main_export(
  File "/home/kahmed10/.local/lib/python3.10/site-packages/optimum/exporters/onnx/__main__.py", line 258, in main_export
    task = TasksManager.infer_task_from_model(model_name_or_path)
  File "/home/kahmed10/.local/lib/python3.10/site-packages/optimum/exporters/tasks.py", line 1764, in infer_task_from_model
    inferred_task_name = cls._infer_task_from_model_name_or_path(
  File "/home/kahmed10/.local/lib/python3.10/site-packages/optimum/exporters/tasks.py", line 1678, in _infer_task_from_model_name_or_path
    library_name = cls.infer_library_from_model(
  File "/home/kahmed10/.local/lib/python3.10/site-packages/optimum/exporters/tasks.py", line 1918, in infer_library_from_model
    library_name = cls._infer_library_from_model_name_or_path(
  File "/home/kahmed10/.local/lib/python3.10/site-packages/optimum/exporters/tasks.py", line 1880, in _infer_library_from_model_name_or_path
    raise ValueError(
ValueError: The library name could not be automatically inferred. If using the command-line, please provide the argument --library {transformers,diffusers,timm,sentence_transformers}. Example: `--library diffusers`.
```

**Comment by IlyasMoutawwakil** [comment]
hi, the pipeline should be loadable with diffusers first, stability has a diffusers-compatible pipeline in stabilityai/stable-diffusion-3-medium-diffusers

**Comment by kahmed10** [comment]
> hi @kahmed10, the pipeline should be loadable with diffusers first, stability has a diffusers-compatible pipeline in [`stabilityai/stable-diffusion-3-medium-diffusers`](https://huggingface.co/stabilityai/stable-diffusion-3-medium-diffusers)

This worked, thanks!!!
```
optimum-cli export onnx --model stabilityai/stable-diffusion-3-medium-diffusers  models/sd3-onnx
```

**Comment by kahmed10** [comment]
@IlyasMoutawwakil is this planning to be merged in soon? I see in CI there are some tests failing.

**Comment by IlyasMoutawwakil** [comment]
@kahmed10 it should be merged today, the failing tests are unrelated (because tests saturating the memory).

**Comment by IlyasMoutawwakil** [comment]
tests are passing except for macos-13 (due to known issue handled in #2085)

File: optimum/exporters/onnx/model_configs.py [hunk 3]
```diff
@@ -1182,6 +1182,97 @@ def outputs(self) -> Dict[str, Dict[int, str]]:
         }
 
 
+class T5EncoderOnnxConfig(CLIPTextOnnxConfig):
```
Comment (echarlaix) [inline optimum/exporters/onnx/model_configs.py:1185]: why is inheriting from `CLIPTextOnnxConfig` needed ?

File: optimum/exporters/onnx/model_configs.py [hunk 4]
```diff
@@ -1037,22 +1044,13 @@ def outputs(self) -> Dict[str, Dict[int, str]]:
...
-        dummy_inputs = super().generate_dummy_inputs(framework=framework, **kwargs)
-
-        # TODO: fix should be by casting inputs during inference and not export
-        if framework == "pt":
-            import torch
-
-            dummy_inputs["input_ids"] = dummy_inputs["input_ids"].to(dtype=torch.int32)
-        return dummy_inputs
-
     def patch_model_for_export(
```
Comment (echarlaix) [inline optimum/exporters/onnx/model_configs.py:1054]: looks like `patch_model_for_export` might also be removed

**Review by echarlaix — APPROVED** [review]
Looks great thanks @IlyasMoutawwakil

File: optimum/utils/input_generators.py [hunk 5]
```diff
@@ -1411,3 +1411,76 @@ def generate(
         float_dtype: str = "fp32",
     ):
         return self.random_int_tensor(shape=(1,), min_value=20, max_value=22, framework=framework, dtype=int_dtype)
+
+
+class DummyTransformerTimestpsInputGenerator(DummyTimestepInputGenerator):
```
[the remark this answers is not in the document]
Comment (IlyasMoutawwakil) [reply optimum/utils/input_generators.py:1416]: it's for the `pipeline.transformer` component

File: optimum/exporters/onnx/model_configs.py (hunk 3 above)
Comment (IlyasMoutawwakil) [reply optimum/exporters/onnx/model_configs.py]: no need, fixed

File: optimum/exporters/onnx/model_configs.py (hunk 4 above)
Comment (IlyasMoutawwakil) [reply optimum/exporters/onnx/model_configs.py:1054]: removing `patch_model_for_export` is handled by https://github.com/huggingface/optimum/pull/2066

File: optimum/utils/input_generators.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (IlyasMoutawwakil) [reply optimum/utils/input_generators.py:1416]: ah okay, I also mispelled `Timestep`, fixing it !

**Comment by IlyasMoutawwakil** [comment]
will merge once all ci is finished !

**Comment by xhcao** [comment]
Hi, @IlyasMoutawwakil Did this patch support to export the series of SD3 to onnx models, include SD3.5 large, SD3.5 Large Turbo, SD3.5 Medium and SD3.0-medium?

**Pull request merged**